### PR TITLE
100 persist irvballotinterpretation

### LINKED ISSUE #100: Persist IRVBallotInterpretation
Implement persistence for IRV Ballot interpretations, based on IRVBallotInterpretation.java in the prototype. Note that you do _not_ need to implement InterpretValidIntent - this has now been moved to IRVChoices.

### PR DESCRIPTION
**Pull request by vteague**
This adds a new database table for persisting the valid interpretations of invalid IRV ballots. This is not actually used in the RLA computation, but is reported in its own report (also part of this PR).

Note this only deals with the UPLOAD Cvrs - the audit CVRs will need to be added, too, but that is in a separate card.

### CODE REVIEW

File: server/eclipse-project/src/main/java/au/org/democracydevelopers/corla/model/GenerateAssertionsSummary.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+  @GeneratedValue(strategy = GenerationType.IDENTITY)
+  private long id;
+
+  /**
+   * Version. Used for optimistic locking.
+   */
+  @Version
+  @Column(name = "version", updatable = false, nullable = false)
+  private long version;
+
```
Comment (michelleblom) [inline server/eclipse-project/src/main/java/au/org/democracydevelopers/corla/model/GenerateAssertionsSummary.java:61]: If this is meant to be a non-updatable data structure, then it is odd to have non final public attributes. I would make them private and just have getters for the required fields.

File: server/eclipse-project/src/main/java/au/org/democracydevelopers/corla/model/vote/IRVBallotInterpretation.java [hunk 2]
```diff
@@ -1,25 +1,146 @@
...
+  @Column(name = "interpretation", columnDefinition = "character varying (1024)", nullable = false)
+  @Convert(converter = StringListConverter.class)
+  private List<String> interpretation = new ArrayList<>();
+
+  /**
+   * Construct an empty IRVBallotInterpretation (for persistence).
+   */
+  public IRVBallotInterpretation() {
+  }
+
```
Comment (michelleblom) [inline server/eclipse-project/src/main/java/au/org/democracydevelopers/corla/model/vote/IRVBallotInterpretation.java:87]: Worth adding in the comments that we are creating a record of an interpretation of an invalid IRV vote.

File: server/eclipse-project/src/main/java/au/org/democracydevelopers/corla/model/vote/IRVBallotInterpretation.java [hunk 3]
```diff
@@ -1,25 +1,146 @@
...
+   * Create a record of an IRV vote interpretation for a given contest and a given ballot
+   * (identified by the CVR ID).
+   * @param contest        the Contest
+   * @param recordType     the type, expected to be either UPLOADED, AUDITOR_ENTERED, or REAUDITED.
+   * @param cvrNumber      the cvr Number, which appears in the csv file (not to be confused with the cvr_id, which the
+   *                       database makes).
+   * @param imprintedId    the imprinted ID, i.e. tabulator_id-batch_id-record_id.
+   * @param rawChoices     the (invalid) raw IRV choices, e.g. [Bob(1),Alice(3),Chuan(4)].
+   * @param orderedChoices the way colorado-rla interpreted the raw choices, as an order list of names.
+   */
```
Comment (michelleblom) [inline server/eclipse-project/src/main/java/au/org/democracydevelopers/corla/model/vote/IRVBallotInterpretation.java:98]: Does line 99 fall within the line length guideline?

File: server/eclipse-project/src/main/java/au/org/democracydevelopers/corla/model/vote/IRVBallotInterpretation.java [hunk 4]
```diff
@@ -1,25 +1,146 @@
...
-The colorado-rla IRV extensions are free software: you can redistribute it and/or modify it under the terms
-of the GNU Affero General Public License as published by the Free Software Foundation, either
-version 3 of the License, or (at your option) any later version.
+/**
+ * A record of the interpretation of invalid IRV ballots. Used for reporting, not for the actual
+ * auditing logic. Every time an invalid IRV vote is encountered, whether in the initial upload or
+ * during the auditing steps, a record is made of the CVR ID, the county, contest name, raw vote,
+ * and our interpretation of that vote as an ordered list of candidate names.
+ * This is output as a report called IRVBallotInterpretationReport.
+ */
```
Comment (michelleblom) [inline server/eclipse-project/src/main/java/au/org/democracydevelopers/corla/model/vote/IRVBallotInterpretation.java:21]: Check whether we need this to implement the Serializable interface in addition to PersistentEntity. If not, some of the methods can be removed if we don't implement it.

File: server/eclipse-project/src/main/java/au/org/democracydevelopers/corla/model/vote/IRVBallotInterpretation.java [hunk 5]
```diff
@@ -1,25 +1,146 @@
...
+   */
+  public IRVBallotInterpretation(Contest contest, CastVoteRecord.RecordType recordType, int cvrNumber, String imprintedId, List<String> rawChoices, List<String> orderedChoices) {
+    this.contest = contest;
+    this.recordType = recordType;
+    this.cvrNumber = cvrNumber;
+    this.imprintedID = imprintedId;
+    this.rawChoices = rawChoices;
+    this.interpretation = orderedChoices;
+  }
+
```
Comment (michelleblom) [inline server/eclipse-project/src/main/java/au/org/democracydevelopers/corla/model/vote/IRVBallotInterpretation.java:107]: "Output details of the stored IRV vote interpretation as a String ..."

File: server/eclipse-project/src/main/java/au/org/democracydevelopers/corla/model/vote/IRVBallotInterpretation.java [hunk 6]
```diff
@@ -1,25 +1,146 @@
...
+  /**
+   * Output contents as a String appropriate for a log message.
+   * @return the data with headers incorporated.
+   */
```
Comment (michelleblom) [inline server/eclipse-project/src/main/java/au/org/democracydevelopers/corla/model/vote/IRVBallotInterpretation.java:111]: I would probably make the input arguments final. (eg. final cvrNumberHeader etc). I don't know that we have followed this convention consistently to catch bugs where input arguments are accidentally reassigned.

File: server/eclipse-project/src/main/java/au/org/democracydevelopers/corla/query/GenerateAssertionsSummaryQueries.java [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+import org.hibernate.Session;
+import us.freeandfair.corla.persistence.Persistence;
+
+import javax.persistence.TypedQuery;
+import java.util.List;
+import java.util.Optional;
+
+import static au.org.democracydevelopers.corla.endpoint.GenerateAssertions.UNKNOWN_WINNER;
+
+/**
```
Comment (michelleblom) [inline server/eclipse-project/src/main/java/au/org/democracydevelopers/corla/query/GenerateAssertionsSummaryQueries.java:36]: "query to retrieve"

File: server/eclipse-project/src/main/java/au/org/democracydevelopers/corla/query/GenerateAssertionsSummaryQueries.java [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+   * @return the (optional) summary of assertions defined for the contest.
+   * @throws RuntimeException when an unexpected error arose in assertion retrieval (not including
+   * a NoResultException, which is handled by returning an empty optional item).
+   */
+  public static Optional<GenerateAssertionsSummary> matching(final String contestName) throws RuntimeException {
+    final String prefix = "[matching]";
+    try {
+      LOGGER.debug(String.format("%s Select query on generate assertions summary table, retrieving " +
+          "summary for contest with name %s.", prefix, contestName));
+
```
Comment (michelleblom) [inline server/eclipse-project/src/main/java/au/org/democracydevelopers/corla/query/GenerateAssertionsSummaryQueries.java:77]: colorado-rla seems to put queries in the "queries" package. I don't know if we want to push the query construction and execution  into AssertionQueries.java?

File: server/eclipse-project/src/main/java/au/org/democracydevelopers/corla/query/GenerateAssertionsSummaryQueries.java [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+      final Session s = Persistence.currentSession();
+      final TypedQuery<GenerateAssertionsSummary> q = s.createQuery("select ca from GenerateAssertionsSummary ca "
+            + " where ca.contestName = :contestName", GenerateAssertionsSummary.class);
+      q.setParameter("contestName", contestName);
+
```
Comment (michelleblom) [inline server/eclipse-project/src/main/java/au/org/democracydevelopers/corla/query/GenerateAssertionsSummaryQueries.java:82]: Is result final?

File: server/eclipse-project/src/main/java/us/freeandfair/corla/csv/DominionCVRExportParser.java [hunk 10]
```diff
@@ -667,21 +668,22 @@ private CastVoteRecord extractCVR(final CSVRecord the_line) {
       if (present) {
         if(isIRV) {
             // If it is IRV, convert it into an ordered list of names (without parentheses), then
-            // store. If it was not valid, make a warning log message and keep going.
+            // store.
```
Comment (michelleblom) [inline server/eclipse-project/src/main/java/us/freeandfair/corla/csv/DominionCVRExportParser.java:671]: Should orderedChoices, irvInterpretation, and msg be final?

File: server/eclipse-project/src/main/java/us/freeandfair/corla/report/ReportRows.java [hunk 11]
```diff
@@ -357,6 +361,9 @@ public static BigDecimal percentage(final BigDecimal num) {
   /**
```
Comment (michelleblom) [inline server/eclipse-project/src/main/java/us/freeandfair/corla/report/ReportRows.java:361]: "For each contest (per row), ..."

File: server/eclipse-project/src/test/java/au/org/democracydevelopers/corla/csv/DominionCVRExportParserTests.java [hunk 12]
```diff
@@ -525,6 +527,25 @@ public void parseBoulder23Succeeds() throws IOException {
         List.of("Yes/For"));
     assertEquals(Objects.requireNonNull(cvr1.contestInfoForContest(boulder302)).choices(),
         List.of("No/Against"));
+
+    // Make the ranked_ballot_interpretation report.
+    final Map<String, String> files = ExportQueries.sqlFiles();
+    final ByteArrayOutputStream os = new ByteArrayOutputStream();
+    String q = files.get("ranked_ballot_interpretation");
+    ExportQueries.csvOut(q, os);
+
```
Comment (michelleblom) [inline server/eclipse-project/src/test/java/au/org/democracydevelopers/corla/csv/DominionCVRExportParserTests.java:536]: Check for whether following violates line length guideline.

File: server/eclipse-project/src/test/java/au/org/democracydevelopers/corla/csv/DominionCVRExportParserTests.java [hunk 13]
```diff
@@ -525,6 +527,25 @@ public void parseBoulder23Succeeds() throws IOException {
         List.of("Yes/For"));
     assertEquals(Objects.requireNonNull(cvr1.contestInfoForContest(boulder302)).choices(),
         List.of("No/Against"));
+
+    // Make the ranked_ballot_interpretation report.
+    final Map<String, String> files = ExportQueries.sqlFiles();
+    final ByteArrayOutputStream os = new ByteArrayOutputStream();
```
Comment (michelleblom) [inline server/eclipse-project/src/test/java/au/org/democracydevelopers/corla/csv/DominionCVRExportParserTests.java:533]: Check whether non-final declared variables can be made final.

File: server/eclipse-project/src/test/java/au/org/democracydevelopers/corla/endpoint/GetAssertionsTests.java [hunk 14]
```diff
@@ -92,6 +102,14 @@ public class GetAssertionsTests {
      */
```
Comment (michelleblom) [inline server/eclipse-project/src/test/java/au/org/democracydevelopers/corla/endpoint/GetAssertionsTests.java:102]: The baseURL attribute is just labelled 'static' without a private/public etc.

File: server/eclipse-project/src/test/java/au/org/democracydevelopers/corla/report/RankedBallotInterpretationReportTests.java [hunk 15]
```diff
@@ -0,0 +1,166 @@
...
+import javax.transaction.Transactional;
+
+import static au.org.democracydevelopers.corla.util.testUtils.*;
+import static org.testng.Assert.*;
+
+import static us.freeandfair.corla.query.CountyQueries.fromString;
+
+/**
+ * Test that a correct ranked vote interpretation report is produced, for a variety of IRV-related
+ * parsing examples, including:
```
Comment (michelleblom) [inline server/eclipse-project/src/test/java/au/org/democracydevelopers/corla/report/RankedBallotInterpretationReportTests.java:54]: will -> with

File: server/eclipse-project/src/test/java/au/org/democracydevelopers/corla/report/RankedBallotInterpretationReportTests.java [hunk 16]
```diff
@@ -0,0 +1,166 @@
...
+ * ThreeCandidatesTenInvalidVotes - a constructed test with invalid IRV ballots to check they are accepted and properly
+ *   interpreted.
+ * An equivalent test using real data from Boulder '23 is in DominionCVRExportParserTests.
+ * Test csvs are a subset of those in DominionCVRExportParserTests.
+ */
+public class RankedBallotInterpretationReportTests extends TestClassWithDatabase {
+
+  /**
+   * Class-wide logger
+   */
```
Comment (michelleblom) [inline server/eclipse-project/src/test/java/au/org/democracydevelopers/corla/report/RankedBallotInterpretationReportTests.java:65]: Generally we don't have the au.org. etc prefix for the Class.class statement here.

File: server/eclipse-project/src/test/java/au/org/democracydevelopers/corla/report/RankedBallotInterpretationReportTests.java [hunk 17]
```diff
@@ -0,0 +1,166 @@
...
+  @Test
+  @Transactional
+  public void parseThreeCandidatesTenInvalidVotesSucceeds() throws IOException {
+    testUtils.log(LOGGER, "parseThreeCandidatesTenInvalidVotesSucceeds");
+    Path path = Paths.get(TINY_CSV_PATH + "ThreeCandidatesTenInvalidVotes.csv");
+    Reader reader = Files.newBufferedReader(path);
+
+    DominionCVRExportParser parser = new DominionCVRExportParser(reader, fromString("Gilpin"), blank, true);
+    assertTrue(parser.parse().success);
+
```
Comment (michelleblom) [inline server/eclipse-project/src/test/java/au/org/democracydevelopers/corla/report/RankedBallotInterpretationReportTests.java:134]: Should q and cvr be final?

**Review by michelleblom — APPROVED** [review]

File: server/eclipse-project/src/main/java/au/org/democracydevelopers/corla/model/vote/IRVBallotInterpretation.java (hunk 4 above)
Comment (vteague) [reply server/eclipse-project/src/main/java/au/org/democracydevelopers/corla/model/vote/IRVBallotInterpretation.java:21]: I got rid of Serizalizable but unfortunately those 3 overridden messages are still needed for PersistentEntity.

File: server/eclipse-project/src/main/java/au/org/democracydevelopers/corla/model/vote/IRVBallotInterpretation.java (hunk 6 above)
Comment (vteague) [reply server/eclipse-project/src/main/java/au/org/democracydevelopers/corla/model/vote/IRVBallotInterpretation.java:111]: Good idea. I've done that for this file and maybe it'd be good to go through the rest too.

**Pull request merged**